pkg/rabbit/connections_handler.go

```diff
@@ -0,0 +1,258 @@
+	GetChannel() RabbitMQChannelInterface
+	SetupRabbitMQConnectionAndChannel(context.Context, string, func(RabbitMQConnectionInterface, RabbitMQChannelInterface) error, func(string) (RabbitMQConnectionWrapperInterface, error))
+}
+
+type RabbitMQConnectionInterface interface {
+	NotifyClose(chan *amqp091.Error) chan *amqp091.Error
+	Close() error
+	IsClosed() bool
+}
+
+type RabbitMQConnectionWrapperInterface interface {
+	RabbitMQConnectionInterface
+	ChannelWrapper() (RabbitMQChannelInterface, error)
+}
+
+type RabbitMQChannelInterface interface {
+	IsClosed() bool
+	NotifyClose(chan *amqp091.Error) chan *amqp091.Error
+	Qos(int, int, bool) error
+	Confirm(bool) error
+	Consume(string, string, bool, bool, bool, bool, amqp091.Table) (<-chan amqp091.Delivery, error)
+	PublishWithDeferredConfirm(string, string, bool, bool, amqp091.Publishing) (*amqp091.DeferredConfirmation, error)
+	QueueInspect(string) (amqp091.Queue, error)
+}
+
+type RabbitMQConnectionHandler struct {
+	firstSetup    bool
+	cycleDuration time.Duration
+	Connection    RabbitMQConnectionWrapperInterface
+	Channel       RabbitMQChannelInterface
+
+	logger *zap.SugaredLogger
+	mu     sync.Mutex
+}
+
+type RabbitMQConnection struct {
+	connection interface{}
+}
+
+func NewConnection(conn interface{}) *RabbitMQConnection {
+	return &RabbitMQConnection{connection: conn}
+}
+
+func (r *RabbitMQConnection) ChannelWrapper() (RabbitMQChannelInterface, error) {
+	if c, ok := r.connection.(*amqp091.Connection); ok {
+		return c.Channel()
+	} else if ci, ok := r.connection.(RabbitMQConnectionWrapperInterface); ok {
+		return ci.ChannelWrapper()
+	}
+	return nil, errors.New("wrong typed connection arg")
+}
+
+func (r *RabbitMQConnection) NotifyClose(c chan *amqp091.Error) chan *amqp091.Error {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.NotifyClose(c)
+	}
+	close(c)
+	return nil
+}
+
+func (r *RabbitMQConnection) Close() error {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.Close()
+	}
+	return errors.New("wrong typed connection arg")
+}
+
+func (r *RabbitMQConnection) IsClosed() bool {
+	if ci, ok := r.connection.(RabbitMQConnectionInterface); ok {
+		return ci.IsClosed()
+	}
+	return true
+}
+
+func NewRabbitMQConnectionHandler(cycleDuration time.Duration, logger *zap.SugaredLogger) RabbitMQConnectionsHandlerInterface {
+	return &RabbitMQConnectionHandler{
+		firstSetup:    true,
+		cycleDuration: cycleDuration,
+		logger:        logger,
+	}
+}
+
+func (r *RabbitMQConnectionHandler) SetupRabbitMQConnectionAndChannel(
+	ctx context.Context,
+	rabbitMQURL string,
+	configFunction func(RabbitMQConnectionInterface, RabbitMQChannelInterface) error,
+	dialFunc func(string) (RabbitMQConnectionWrapperInterface, error)) {
+	var err error
+	wait := true
+	r.mu.Lock()
+	if r.firstSetup {
+		wait = false
+	}
+	r.mu.Unlock()
+	r.logger.Info("Creating and Configuring RabbitMQ Connection and Channel")
+	for {
+		if wait {
```

Comment:
nit: I think we can just use the single firstSetup bool instead of tracking wait separately